lib/demuxeres/gstdemuxeres.c

```diff
@@ -0,0 +1,655 @@
+    default:
+      break;
+  }
+  g_free (stream);
+}
+
+GstDemuxerEStream *
+gst_parse_stream_create (GstDemuxerES * demuxer, GstPad * pad)
+{
+  GstDemuxerEStream *stream;
+  GstElement *appsink = NULL;
+  GstAppSinkCallbacks callbacks = { 0, };
+  GstPad *sinkpad;
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  GstPadLinkReturn link;
+  GstCaps *caps = gst_pad_query_caps (pad, NULL);
+
+  if (!caps) {
+    GST_ERROR
+        ("Unable to get the caps from pad, unable to create a new stream");
+    return NULL;
+  }
+  GST_INFO ("the stream caps is %" GST_PTR_FORMAT, caps);
+
+  stream = g_new0 (GstDemuxerEStream, 1);
+  stream->demuxer = demuxer;
+  stream->type = gst_parse_stream_get_type_from_pad (pad);
+
+  GstStructure *s = gst_caps_get_structure (caps, 0);
+  switch (stream->type) {
+    case DEMUXER_ES_STREAM_TYPE_VIDEO:
+    {
+      gst_video_info_from_caps(&stream->data.video.info, caps);
+     gst_structure_get_int (s, "bitrate", &stream->data.video.bitrate);
+      stream->data.video.profile =
+          g_strdup (gst_structure_get_string (s, "profile"));
+      stream->data.video.level =
+          g_strdup (gst_structure_get_string (s, "level"));
+      stream->data.video.vcodec = gst_parse_stream_get_vcodec_from_caps (caps);
+      break;
+    }
+    case DEMUXER_ES_STREAM_TYPE_AUDIO:
+    {
+      gst_audio_info_from_caps(&stream->data.audio.info, caps);
+      gst_structure_get_int (s, "bitrate", &stream->data.audio.bitrate);
+      stream->data.audio.acodec = gst_parse_stream_get_acodec_from_caps (caps);
+      break;
+    }
+    default:
+      break;
+  }
+  stream->id = g_list_length (priv->streams);
+
+  appsink = gst_element_factory_make ("appsink", NULL);
+  g_object_set (appsink, "sync", FALSE, "emit-signals", TRUE, NULL);
+
+  gst_bin_add_many (GST_BIN (priv->pipeline), appsink, NULL);
+  sinkpad = gst_element_get_static_pad (appsink, "sink");
+  if ((link = gst_pad_link (pad, sinkpad)) != GST_PAD_LINK_OK) {
+    GST_ERROR ("Unable to link the appsink with the parser link status %d",
+        link);
+    gst_parse_stream_teardown(stream);
+    stream = NULL;
+  }
+
+  GST_DEBUG_BIN_TO_DOT_FILE_WITH_TS (GST_BIN (priv->pipeline),
+      GST_DEBUG_GRAPH_SHOW_ALL, "gst-demuxerer.stream_create");
+
+  if (appsink)
+    gst_element_sync_state_with_parent (appsink);
+
+  callbacks.new_sample =appsink_new_sample_cb;
+  gst_app_sink_set_callbacks ((GstAppSink *) appsink, &callbacks, stream,
+      NULL);
+
+  gst_caps_unref (caps);
+  gst_object_unref (sinkpad);
+  return stream;
+}
+
+static inline gboolean
+wait_for_demuxer_ready (GstDemuxerES * demuxer, gint64 wait_time)
+{
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  gboolean success = FALSE;
+  gint64 end_time = g_get_monotonic_time () + wait_time;
+
+  g_mutex_lock (&priv->ready_mutex);
+  while (!success) {
+
+    success = g_cond_wait_until (&priv->ready_cond,
+        &priv->ready_mutex, end_time);
+    GST_DEBUG ("After wait success %d", success);
```

Comment:
according to documentation https://docs.gtk.org/glib/method.Cond.wait_until.html this doesn't look complete, since the *predicate* must be check to break the loop, and in this case, the predicate is if the number of streams in `priv->streams` has increased.